Syntax: Don't highlight the LHS of type decl as types

Julow opened the pull request:
I would expect `ocamlTypeConstr` to apply only to type constructors
within type expressions, not to the identifier after `type` in:

```ocaml
type foo = 'a bar list
```

`foo` is not highlighted with:
```viml
hi link ocamlTypeConstr Type
```

This makes the LHS of types, exceptions and constraints be matched as
`ocamlTypedef`, which is not highlighted by default.

Variance and type variables on the LHS remain highlighted as before.

Julow commented:
Ping @Maelan

Maelan commented:
Thanks @Julow! I’m fine either way. I’m just slightly concerned by adding complexity to that already complex piece, but I let our maintainers judge.

You’d also have to deal with `type t = ... and u = ...`, though.

Julow commented on syntax/ocaml.vim:
```diff
@@ -465,13 +465,23 @@ syn region ocamlTypeSumAnnot contained
-\ matchgroup=NONE end="\(\<type\>\|\<exception\>\|\<val\>\|\<module\>\|\<class\>\|\<method\>\|\<constraint\>\|\<inherit\>\|\<object\>\|\<struct\>\|\<open\>\|\<include\>\|\<let\>\|\<external\>\|\<in\>\|\<end\>\|)\|]\|}\|;\|;;\)\@="
-\ contains=@ocamlTypeExpr,ocamlTypeEq,ocamlTypePrivate,ocamlTypeDefDots,ocamlTypeRecordDecl,ocamlTypeSumDecl,ocamlTypeDefAnd,ocamlComment,ocamlPpx
-hi link ocamlTypeDef ocamlTypeCatchAll
+\ matchgroup=ocamlTypedef end="\<\(\w\|'\)*\>"
```
This should actually be one of the "identifier" group. I've splitted the exceptions out, this is now `ocamlLCIdentifier` and exceptions now use `ocamlConstructor`, as before.

copy replied:
I don't think `ocamlLCIdentifier` is a good choice (even if technically correct): It makes it impossible to highlight these `foo`s  differently, even though they are in different namespaces:

```ocaml
let foo = ...
type foo = ...
```

Julow replied:
I added a new match group. This restores the highlighting you want:
```
hi link ocamlTypeIdentifier ocamlTypeConstr
```

Julow commented on syntax/ocaml.vim:
```diff
@@ -465,13 +465,23 @@ syn region ocamlTypeSumAnnot contained
+\ matchgroup=ocamlTypedef end="\<\(\w\|'\)*\>"
+\ contains=@ocamlAllErrs,ocamlComment,ocamlTypeVariance,ocamlTypeVar
+\ skipwhite skipempty
+\ nextgroup=ocamlTypeDefImpl
```
I don't see this. Perhaps because the `end` regex don't contain `\@=` ?

Maelan replied:
I haven’t actually tested your code, but from what I understand about `nextgroup=` and `contained`, the `ocamlTypeDefImpl` group will never be tried except immediately after the end of the `ocamlTypeDef` group.

Julow replied:
This is what `skipwhite` is doing, from the `:h`:
```
These arguments are only used in combination with "nextgroup".	They can be
used to allow the next group to match after skipping some text:
	skipwhite	skip over space and tab characters
	skipnl		skip over the end of a line
	skipempty	skip over empty lines (implies a "skipnl")
```

I took thought this was doing something between `start` and `end` and nothing outside but it's actually the opposite, they take effect after `end` and do nothing inside the region (whitespace are skipped by default it seems).

Julow commented:
I fixed the problems with `and` and restored the highlighting of `exception`. `module-rec-and` (which are broken on master) and `let-and` are not altered.

Comparison before/after with `hi link ocamlTypeConstr Type`:

![image](https://user-images.githubusercontent.com/2310568/179964354-3f61683d-b89a-4bf3-b912-9d48d4712c9c.jpg)

copy commented:
Fundamentally this change is fine, but I think we leave the current behavior configurable (personally, I prefer the current behavior).

This seems to break some things, in particular (from type-linter-test.ml):

```ocaml
  type ('a, 'b) t = 'a list * ('a, 'b) result (* the left pair of parenthesis *)
  type nonrec 'a o = 'a option = private None | Some of 'a (* the second equal sign *)
  type t += C of int (* the of *)
```

The module-rec-and find is interesting, would you mind adding it to type-linter-test.ml with a fixme?

rgrinberg commented on syntax/ocaml.vim:
```diff
@@ -6,6 +6,7 @@
 "               Issac Trotts      <ijtrotts@ucdavis.edu>
 " URL:          https://github.com/ocaml/vim-ocaml
 " Last Change:
+"               2022 Jul 20 - Improved highlighting of type decl (Jules Aguillon)
```
As a side note, the previous date: 2019 isn't really correct.

Julow replied:
It seems to really be authored in 2019 but committed in 2022: 41fb0d069ccdb52a305a5df5b515ff43bb4c925e

Maelan replied:
Agreed with @rgrinberg. I did most of the work back in 2019 (not everything though: the recent polishing work got squashed into that commit, which kept the earlier date), and I thought it was relevant to mention that date (because of the evolution of the language), *but* I realize now that people would rather be interested in the release date.

Julow replied:
I added a commit for the last two changes.

Julow commented:
I missed type-linter-test.ml, thanks.

I've added a way to get back the current behavior, what should be the default ? This is governed by `hi link ocamlTypeIdentifier x` which could be `ocamlLCIdentifier` (new) or `ocamlTypeConstr` (previous).

I've fixed the bugs with type params and equal signs but I introduced new bugs around let-bindings, which are matched as type decl. I would like `ocamlTypeDefImpl` to take effect only when invoked with `nextgroup`, is here a way to do that ? Otherwise, we need stronger rules for let-bindings.

Julow commented:
Any opinion on this ?

Here's how it can be used:

```viml
hi link ocamlTypeConstr Type
hi link ocamlTypeIdentifier Identifier
```

The first line is both before and after this PR. The second line is with the two line above applied:

![vimocamlshot](https://github.com/ocaml/vim-ocaml/assets/2310568/1add7307-78f4-4880-8d4f-c7a3613ce60b)

rgrinberg commented:
@copy do you want to make the call on this?

copy commented:
This change and the mechanism to configure it looks good.

However, it causes some problems with the highlighting of number literals (highlighted as error across the [type-linter-test.ml](https://github.com/ocaml/vim-ocaml/blob/master/type-linter-test.ml) file), the `fun` keyword (highlighted like a type on line 332 and 338) and comments (a correct type highlighting is removed from line 349). You can use vim's `:TOhtml` followed by diff (or vimdiff, patdiff, ...) to check.

Julow commented:
Thanks for the review! I fixed this issue, it was a missing `ocamlContained` declaration. I also changed my editor setup to use this patch everyday to be sure to test it well.

copy commented:
Thanks! Super-minor, but it still removes the type highlighting from the following (on the int):

```ocaml
  type t (*c*) =  int
```

Julow commented:
Good catch! I'm failing to fix it.

The `ocamlTypeDef` region ends at `t` and jump into the `ocamlTypeDefImpl` region, which starts at `=` using `nextgroup`. I couldn't find a way to accept the comment without interfering with `nextgroup`, which is very handy to handle types without RHS and `and` types.
Any idea ?

copy commented:
This comment from @Maelan looks relevant: https://github.com/ocaml/vim-ocaml/pull/81#discussion_r924975501

Julow commented:
Sorry for the slow reply. I managed to handle the comment at this location while keeping `nextgroup` working for `and` types.

copy commented:
The latest commit seems to have broken the highlighting of identifiers in some places (e.g. the `f` in line 49 is recognized as a type). You can use `:TOhtml` and diff to find any differences between this branch and master.

Julow commented:
I think I managed to fix the identifier issue. The TOhtml diff seems fine now. The added `Error` class is due to a newly added test case.

copy commented:
On the last commit (817e8f75a8d1324b2d28c0958534a742656d8350), some highlights are still incorrect (e.g. the `let f` on line 49), and show up in the diff.
